lib/datadog/core/telemetry/collector.rb

```diff
@@ -0,0 +1,241 @@
+            kernel_name: Core::Environment::Platform.kernel_name,
+            kernel_release: Core::Environment::Platform.kernel_release,
+            kernel_version: Core::Environment::Platform.kernel_version
+          )
+        end
+
+        # Forms a telemetry app-started integrations object
+        def integrations
+          Datadog.registry.map do |integration|
+            is_instrumented = instrumented?(integration)
+            is_enabled = is_instrumented && patched?(integration)
+            Telemetry::V1::Integration
+              .new(
+                name: integration.name.to_s,
+                enabled: is_enabled,
+                version: integration_version(integration),
+                compatible: integration_compatible?(integration),
+                error: (patch_error(integration) if is_instrumented && !is_enabled),
+                auto_enabled: is_enabled ? integration_auto_instrument?(integration) : nil
+              )
+          end
+        end
+
+        # Returns the runtime ID of the current process
+        def runtime_id
+          Datadog::Core::Environment::Identity.id
+        end
+
+        # Returns the current as a UNIX timestamp in seconds
+        def tracer_time
+          Time.now.to_i
+        end
+
+        private
+
+        def bundled_environment?
+          begin
+            !Bundler.bundle_path.nil?
+          rescue Bundler::GemfileNotFound
+            false
+          end
+        end
+
+        def env
+          Datadog.configuration.env
+        end
+
+        def service_name
+          Datadog.configuration.service
+        end
+
+        def service_version
+          Datadog.configuration.version
+        end
+
+        def tracer_version
+          Core::Environment::Identity.tracer_version
+        end
+
+        def products
+          profiler_obj = profiler
+          appsec_obj = appsec
+          profiler_obj || appsec_obj ? Telemetry::V1::Product.new(profiler: profiler_obj, appsec: appsec_obj) : nil
+        end
+
+        def profiler
+          version = profiler_version
+          Telemetry::V1::Profiler.new(version: version) if version
+        end
+
+        def profiler_version
+          tracer_version if Datadog.configuration.respond_to?(:profiling) && Datadog.configuration.profiling.enabled
+        end
+
+        def appsec
+          version = appsec_version
+          Telemetry::V1::AppSec.new(version: version) if version
+        end
+
+        def appsec_version
+          tracer_version if Datadog.configuration.respond_to?(:appsec) && Datadog.configuration.appsec.enabled
+        end
+
+        def configuration_variables
+          configurations = []
+          environment_collector = Core::Diagnostics::EnvironmentCollector.new
+          [
+            Telemetry::V1::Configuration.new(
+              name: 'DD_AGENT_HOST', value: Datadog.configuration.agent.host || ENV.fetch('DD_AGENT_HOST', '127.0.0.1')
```

Comment:
I recommend using the hostname from: https://github.com/DataDog/dd-trace-rb/blob/265fa48a736527241265c07a768f9883679eed30/lib/ddtrace/transport/http/adapters/net.rb#L80-L82

As this will be definitely the correctly configured value, and will always match the value returned by `environment_collector.agent_url` a few lines down.

Reply:
You might have to expose a few more methods, but I think that's worth it.